### ISSUE TITLE
Wrong 3D chart height / title placement if legend position left/right

### ISSUE THREAD

**Issue by lticadler** [amcharts/amcharts4#3604]
**Bug description**

Wrong 3D chart height / title placement if legend position left/right.
Two times the same chart - but different legend position:
![image](https://user-images.githubusercontent.com/58069494/122872105-3e2d9e00-d330-11eb-8df6-c0610dc3594b.png)

Codepen: https://codepen.io/chrisadler/pen/wvJbwdN?editors=1010

**Environment (if applicable)**

 - amCharts version: 4.10.20 (and previous versions)
 - Browser and its version: all Browsers

**Comment by martynasma** [comment]
I'd try to move to a `topAxesContainer`:

```JavaScript
const title = chart.titles.create();
title.text = "Title"+no;
title.fontSize = 25;
title.parent = chart.topAxesContainer;
```

**Comment by lticadler** [comment]
Unfortunately setting title parent to topAxesContainer doesn't really help.

With just the title parent set both charts have the same problem:
![image](https://user-images.githubusercontent.com/58069494/123032597-7f808500-d3e6-11eb-9239-6d888fd494eb.png)
Of course, one could say set the title's marginBottom - but the font size is chosen by the customer - therefore there is no fix solution.

And even worse - if the scrollbar is active the title is below the scrollbar:
![image](https://user-images.githubusercontent.com/58069494/123032859-06356200-d3e7-11eb-9155-c51567dfd4cf.png)

**Comment by lticadler** [comment]
For comparison I updated the pen with multiple charts. https://codepen.io/chrisadler/pen/wvJbwdN
For the 3D chart the topAxesContainer is no solution.
In common the behavior is different to the 2D chart in case the legend is positioned left (or right) and title.parent = topAxesContainer is even worse :(

![image](https://user-images.githubusercontent.com/58069494/123036579-9a0a2c80-d3ed-11eb-98e6-f1eb5615e710.png)

![image](https://user-images.githubusercontent.com/58069494/123036688-be660900-d3ed-11eb-9af4-9e7e460f6332.png)

![image](https://user-images.githubusercontent.com/58069494/123036720-caea6180-d3ed-11eb-9521-208ffdcbbb4a.png)

![image](https://user-images.githubusercontent.com/58069494/123036767-dccc0480-d3ed-11eb-921c-82a3d7a2c607.png)

![image](https://user-images.githubusercontent.com/58069494/123036882-09801c00-d3ee-11eb-8ab8-f1640261bdf5.png)

**Comment by martynasma** [comment]
Maybe add it to `chartContainer` and not push it to `titles` at all:

```JavaScript
const title = new am4core.Label(); //chart.titles.create();
title.text = `Title${no} 3d:${use3D?true:false} legend:${legendLeft?'left':'bottom'} title-parent:${useTitleParent?true:false} scrollbar:${useScrollbar?true:false}`;
title.fontSize = 14;
title.fill = am4core.color('#0070ff');
title.align = "center"
title.marginBottom = 25;
title.parent = chart.chartContainer;
title.toBack();

// For 3D adjust via dy
title.dy = -25;
```

**Comment by lticadler** [comment]
Unfortunately the chartContainer is as well no solution and causes many other problems.

- Fontsize is adjustable by our customers - therefore the dy would be a problem - even more as angle and depth of the chart are as well set by the customer.
- our customers using multiple titles with different font sizes - impossible to position them all in the same way amCharts does it on it's own in a 2D XY chart (so far we use `chart.titles.push(new am4core.Label());`).

I made a fork of my first pen and set title parent to chartContainer: https://codepen.io/chrisadler/pen/gOmNwwr?editors=0010
![image](https://user-images.githubusercontent.com/58069494/123062670-09434900-d40d-11eb-9f7e-518d64940deb.png)

In my opinion it's a height calculation bug in the XYChart3D if legend is positioned left or right (as we saw in the first pen it works if the legend is on bottom or top).
(Even more as in our case the legend is rendered to it's own div and is not part of the chart container div.)

**Comment by martynasma** [comment]
OK, another suggestion. Use the title as before. Just adjust its `dx` by the width of the legend. E.g.:

```JavaScript
if (legendLeft) {
  chart.events.on("sizechanged", function(ev) {
    var chart = ev.target;
    var title = chart.titles.getIndex(0);
    title.dx = chart.legend.measuredWidth;
  })
}
```

**Comment by lticadler** [comment]
Sadly this doesn't solve the main issue.
The problem is not the x position of the title. The y position is not on top of the chart (especially if a scrollbar is added).

(As mentioned earlier - the legend (in our product) is not inside the chart div - therefore the legend's width doesn't matter. The title is in our case always centered.)

**Comment by lticadler** [comment]
The "legend and chart" area should stay the same - no matter, where the legend is.
(Because of changing the legend's position there is not more space in common for chart legend and title. Div size is still the same.)
![image](https://user-images.githubusercontent.com/58069494/123244624-652cd100-d4e4-11eb-86c9-be187186285a.png)
Forked pen: https://codepen.io/chrisadler/pen/abJeoWO?editors=1010

**Comment by lticadler** [comment]
I changed the pen https://codepen.io/chrisadler/pen/abJeoWO?editors=0010 and added for each chart the same chart but in 2D. 2D is always fine - I would expect 3D to behave the same.
And I made the title longer to see if the title is above the legend even if legend is left/right.
The comparison shows very well the different behavior of the chart's height in case of 3D and legend position left/right.
![image](https://user-images.githubusercontent.com/58069494/123367681-a8d21a00-d57a-11eb-8216-75622fa0d633.png)
![image](https://user-images.githubusercontent.com/58069494/123367709-b5ef0900-d57a-11eb-977e-1546fdf14c30.png)
![image](https://user-images.githubusercontent.com/58069494/123367733-c1423480-d57a-11eb-840f-e3b224ff09e9.png)
![image](https://user-images.githubusercontent.com/58069494/123367765-cc956000-d57a-11eb-92e2-ee7f4df83361.png)

**Comment by martynasma** [comment]
Fixed in [4.10.21(https://github.com/amcharts/amcharts4/releases/tag/4.10.21).

## [4.10.21] - 2021-08-08

### Added
- Export: new setting in CSV options: `addBOM` (default `true`).

### Changed
- `userClassName` now supports multiple class names separated by space.
- Exporting CSV will now have add BOM code to exported file.
- Upgrading to `pdfmake 0.2.2` to work around security issues in older version.
- Upgrading to `xlsx 0.17.0` to work around security issues in older version.

### Fixed
- In some cases axis labels could be hidden across multiple data updates.
- Fixed issue with stacked/clustered negative values.
- `MapChart` now respects `home*` zoom/position with externally loaded geodata.
- Sizing/positioning of 3D charts with titles fixed.
- RangeSelector plugin: fixed zoom range resetting issues.
- Issue with inverted radial axis fixed.
- Axis with `snapToSeries` set was snapping to hidden series.
- Using timezone was causing `DateAxis` tooltips to show incorrect time in some cases.
- Issue with label text with a lot of in-line formatting not wrapping properly fixed.

[Full change log](https://github.com/amcharts/amcharts4/blob/master/dist/es2015/CHANGELOG.md).

[Download options](https://www.amcharts.com/download/).

Make sure you clear your browser cache after upgrading. And feel free to contact us again if you are still experiencing this issue.
